Add signout and 'save and sign out' buttons

insacuri opened the pull request:
### What is the context of this PR?
https://trello.com/c/39EwmSSe/2385-sign-out-of-eq-and-ras

Adds a 'sign out' or 'save and sign out' button to blue banner and removed it from the submit button on the page.

### How to review 
Open a survey and sure that for the introduction page and thank you page, the button in the blue banner reads 'Sign out' and every other page should have 'Save and Sign out' as the button in the blue banner.

Additionally, make the browser smaller to trigger the mobile version of the site and it should appear in the dropdown menu also.
### Checklist

* [ ] New static content marked up for translation
* [ ] Newly defined schema content included in eq-translations repo

insacuri commented:
Adding in progress label because I forgot to run the tests and a bunch of the broke

MebinAbraham commented on app/templates/partials/header.html:
```diff
@@ -24,6 +24,13 @@
                 </div>
+                {% if account_service_log_out_url %}
+                    {% if not content or content.block and content.block.type == 'Introduction' %}
+                    <div class="u-fr u-d-no--@xs@s"><button class="btn btn--ghost print__hidden" data-qa="btn-save-sign-out" type="submit" name="action[save_sign_out]">{{ _("Sign out") }}</button></div>
```
`data-qa` and `action` needs updating?. Also worth adding event for analytics?.

insacuri replied:
Agreed, I did all 3 bits in the end

MebinAbraham reviewed (changes requested):
Layout issue:
<img width="360" alt="screenshot 2018-11-05 at 12 37 05" src="https://user-images.githubusercontent.com/35296336/47998539-910ea480-e0f7-11e8-9b28-6657d2b62bbd.png">

MebinAbraham reviewed (changes requested):
Something odd is happening, when you hover over an answer and hit the `Enter` button, it invokes the save and signout button?
Seems to be doing it for like every answer type.


<img width="583" alt="screenshot 2018-11-05 at 12 40 54" src="https://user-images.githubusercontent.com/35296336/47998692-290c8e00-e0f8-11e8-81cb-e27b7735761c.png">


Form Data:
```
action[save_sign_out]: 
csrf_token: IjBiNTY1NzVhMjE3MjQ3ZWRmMmM5OGM1NTRjOTI1YjdmZGRhNjhiYWMi.DsHNkw.6Vcxidf0ErDOsTjphli1RduIInU
weekly-pay-paid-employees-answer: 9
```

MebinAbraham reviewed (commented):
Does this not need to apply for our other themes?

MebinAbraham commented on app/templates/partials/header.html:
```diff
@@ -24,6 +24,13 @@
                     <h1 class="header__title ">{{ survey_title }}</h1>
                 </div>
+                {% if account_service_log_out_url %}
+                    {% if not content or content.block and content.block.type == 'Introduction' %}
```
I found this kinda confusing, can we make use parentheses?.
`{% if not content or (content.block and content.block.type == 'Introduction') %}`

insacuri replied:
Yea agreed, I added brackets

MebinAbraham reviewed (commented):
The layout column seems to slightly off?
<img width="505" alt="screenshot 2018-11-05 at 13 52 25" src="https://user-images.githubusercontent.com/35296336/48002026-3a5a9800-e102-11e8-8989-10b80538bb0f.png">
According to pattern lib, it should be like this?
<img width="505" alt="screenshot 2018-11-05 at 13 54 32" src="https://user-images.githubusercontent.com/35296336/48002064-59f1c080-e102-11e8-91aa-e8336a898efc.png">

insacuri commented:
@MebinAbraham I've fixed the column layout (didn't realise the header had a few new classes against it).
I also sorted out the alignment issue too.
I've spoken to Matt about pressing enter logging you out and he's going to look into it.

MebinAbraham commented:
My last comment is still an issue when the width is between 500-575px

MebinAbraham reviewed (commented):
I get a 500 error when I click `Sign out` after submission?
```
File "/Users/abrahm/.virtualenvs/eq-survey-runner-TQNVnq61/lib/python3.6/site-packages/werkzeug/routing.py", line 1580, in match
    raise MethodNotAllowed(valid_methods=list(have_match_for))
werkzeug.exceptions.MethodNotAllowed: 405 Method Not Allowed: The method is not allowed for the requested URL.
127.0.0.1 - - [06/Nov/2018 10:42:30] "POST /questionnaire/test/view_submitted_response/thank-you HTTP/1.1" 500 -
```

dwyeradam commented:
Lewis has confirmed now 👍 so ignore below.

@insacuri can you please chase up RAS/RH regarding https://github.com/ONSdigital/ons-schema-definitions/pull/39. Until we confirm that they are going to pass the `account_service_log_out_url` we shouldn't really merge this. That said i'll still review this on the assumption that the above PR is fine.

insacuri commented:
@dwyeradam done, Lewis said it's all good.  Is that enough or do we need another from that team looking at it?

dwyeradam reviewed (dismissed):
At the moment if `account_service_log_out_url` isn't passed we don't have a save and sign out button. Shouldn't we still have the button but if the `account_service_log_out_url` isn't available we go to the current signed-out.html page? It seems to do this for social. Otherwise we're stopping respondents from saving and signing out part way through.

insacuri commented:
@dwyeradam This was a tricky one.  For business surveys we 'should' always have it.  This field is a weird one as it's optional but expected for certain surveys.  Is the correct thing here to have a better fallback option? or can we make that field mandatory for surveys we **know** will have an account service driving the whole thing?

insacuri commented:
@dwyeradam Removed the conditional 'save and sign out' and it works fine.  Should be good to review now

bitdivision commented on tests/functional/base_pages/theme/northernireland/introduction-page.page.js:
```diff
@@ -0,0 +1,11 @@
+// >>> WARNING THIS PAGE WAS AUTO-GENERATED - DO NOT EDIT!!! <<<
+const QuestionPage = require('../../question.page');
```
Don't think these pages should be here?

insacuri replied:
Removed them now

MebinAbraham commented:
Any update from Matt?. Sign out on `Enter` still occurs.

MebinAbraham reviewed (changes requested):
If I go to the feedback page (`/feedback`) then click 'Save and sign out`, instead of signing out, it actually submits the form and displays:
```
Submission successful
Thank you for your feedback.

It is now safe to close this window.
```

---

Also if you then click `Sign out`, you get a 500 error.
```
werkzeug.exceptions.MethodNotAllowed: 405 Method Not Allowed: The method is not allowed for the requested URL.
127.0.0.1 - - [12/Nov/2018 12:50:55] "POST /feedback/thank-you HTTP/1.1" 500 -
```

ajmaddaford commented:
@MebinAbraham @dwyeradam @bitdivision this is ready for re-review.

The changes I've made are:
- Introduced a new top level variable to the templates - `save_on_signout` - this is used to determine if the page is saveable before signing out.
- Refactored the sign out button logic into its own template and added macros for the buttons.
- The CSRF tokens are injected directly in the template rather than via the view context. There was a lot of duplication here previously.
- Added additional functional tests to cover the sign out buttons available on the feedback screen and view submission.

I've raised some issues for further refactoring that I felt would have bloated this pull request - https://github.com/ONSdigital/eq-survey-runner/issues/1918 and https://github.com/ONSdigital/eq-survey-runner/issues/1919.

bitdivision commented on tests/functional/helpers.js:
```diff
@@ -20,8 +22,8 @@ const startCensusQuestionnaire = (schema, sexualIdentity = false, region = 'GB-E
 
-function openQuestionnaire(schema, userId = getRandomString(10), collectionId = getRandomString(10), periodId = '201605', periodStr = 'May 2016', region = 'GB-ENG', language = 'en') {
-  return generateToken(schema, userId, collectionId, periodId, periodStr, region, language)
+function openQuestionnaire(schema, { userId = getRandomString(10), collectionId = getRandomString(10), periodId = '201605', periodStr = 'May 2016', region = 'GB-ENG', language = 'en', sexualIdentity = false, includeLogoutUrl = false } = {}) {
```
:+1:

MebinAbraham commented on data/en/test_theme_northernireland.json:
```diff
@@ -0,0 +1,60 @@
+    "data_version": "0.0.2",
+    "survey_id": "0",
+    "title": "Test Northern Ireland Survey",
+    "theme": "social",
```
This should be `northernireland`

MebinAbraham commented on data/en/test_theme_northernireland.json:
```diff
@@ -0,0 +1,60 @@
+                "id": "summary"
+            }],
+            "id": "group",
+            "title": "Norther Ireland Theme Test"
```
An `n` was lost in transition...

MebinAbraham reviewed (changes requested):
`Sign out` button on the`/feedback` page is submitting the form.

MebinAbraham reviewed (commented):
@ajmaddaford You should also update the message.pot

MebinAbraham commented on app/templates/layouts/_base.html:
```diff
@@ -79,28 +80,31 @@
         </div>
+        <form class="form qa-questionnaire-form" role="form" method="POST" autocomplete="off" novalidate>
+          <input id="csrf_token" name="csrf_token" type="hidden" value="{{ csrf_token() }}"/>
+          <button class="u-vh" tabindex="-1" type="submit" name="action[save_continue]">Save and continue</button>
```
Needs to be marked for translation.

MebinAbraham commented on app/themes/census/templates/introduction.html:
```diff
@@ -24,10 +24,7 @@ <h4 class="venus"> Enter your 20 digit unique access code </h4>
-    {{ content.csrf_token }}
-    <button class="btn qa-btn-get-started" type="submit" name="action[start_questionnaire]">Start survey</button>
-  </form>
+  <button class="btn qa-btn-get-started" type="submit" name="action[start_questionnaire]">Start survey</button>
```
Needs marking for translation.

MebinAbraham replied:
ah its census...

MebinAbraham commented on tests/functional/spec/checkbox.spec.js:
```diff
@@ -93,7 +93,7 @@ describe('Checkbox with "other" option', function() {
   });
 
-  it('Given I have previously added text in other texfiled and saved, when I uncheck other options and select a different checkbox as answer, then the text entered in other field must be wiped.', function() {
+  it('Given I have previously added text in other textfield and saved, when I uncheck other options and select a different checkbox as answer, then the text entered in other field must be wiped.', function() {
```
👍

MebinAbraham reviewed (commented):
Code changes looks good, just that translation comment to fix. Will approve tomorrow morning after a final check.

MebinAbraham commented on app/templates/signed-out.html:
```diff
@@ -4,12 +4,11 @@
 </div>
 
-{% if account_service_url %}
+ {% if account_service_url %}	
```
&nbsp; Dodgy &nbsp; &nbsp; indentation.

MebinAbraham commented on app/views/session.py:
```diff
@@ -72,6 +72,9 @@ def login():
         cookie_session['account_service_url'] = claims.get('account_service_url')
 
+    if 'account_service_log_out_url' in claims and claims.get('account_service_log_out_url'):
+        cookie_session['account_service_log_out_url'] = claims.get('account_service_log_out_url')
```
I know the existing is done like this but can it not just be:
```
 if claims.get('account_service_log_out_url'):
        cookie_session['account_service_log_out_url'] = claims.get('account_service_log_out_url')

MebinAbraham commented on tests/functional/spec/save_sign_out.spec.js:
```diff
@@ -37,4 +52,108 @@ describe('SaveSignOut', function() {
+      });
+  });
+
+  it('@watch  Given I load the feedback page and I sign out, then I should be on the account service logged out page', function() {
```
`@watch`

MebinAbraham commented on tests/functional/spec/save_sign_out.spec.js:
```diff
@@ -3,27 +3,42 @@ const SetMinMax = require('../generated_pages/numbers/set-min-max-block.page.js'
-  it('Given i am returning to the questionnaire, then I am returned to the page I was on and can then complete the survey', function() {
-    return helpers.openQuestionnaire('test_numbers.json', 'test_user', collectionId)
-      .then(() => {
+  it('Given I am returning to the questionnaire, then I am returned to the page I was on and can then complete the survey', function() {
```
Subjective I guess, but we should be doing `Given, When, Then`?

ajmaddaford replied:
Fair point, I've updated all scenarios to be Given, When, Then

MebinAbraham reviewed (approved):
Nice 👍

bitdivision reviewed (commented):
I've looked through all the code :+1:

I've done some testing, but will finish testing and have a last look on Monday morning.

bitdivision commented:
@ajmaddaford 	Not a big deal, but:

![how frequently are employees paid - monthly wages and salaries survey](https://user-images.githubusercontent.com/205209/49372167-b04a2300-f6f1-11e8-812a-024bc3744c66.png)

The save and signout button is not on the same row as the title.

Ah, @MebinAbraham has actually already raised this.

MebinAbraham reviewed (commented):
@ajmaddaford To fix the header issue you just need to edit:
https://github.com/ONSdigital/eq-survey-runner/blob/1bbb7ea0fef9361d0e0982f9f752a10770e2c489/app/themes/northernireland/templates/partials/header.html#L21
https://github.com/ONSdigital/eq-survey-runner/blob/1bbb7ea0fef9361d0e0982f9f752a10770e2c489/app/templates/partials/header.html#L24

Class is missing: `col-6@s`
So should be: 
``` HTML
<div class="grid__col {{ 'col-10@xs col-6@s col-8@m' if account_service_url }}">
```

ajmaddaford committed:
Add signout and 'save and sign out' buttons

- Adds a visually hidden submit button so 'enter' continues rather than
  signs out.
- New top level variable to the templates - save_on_signout - this is
  used to determine if the page is saveable before signing out.
- CSRF tokens are injected directly in the template rather than via the
  view context.
